File: backend/node-service/services/serviceAuth.js
const crypto = require('crypto');
const config = require('./config');

const SIGNATURE_VERSION = 'v1';

function canonicalServiceRequest(method, pathAndQuery, timestamp, nonce, body = Buffer.alloc(0)) {
  const rawBody = Buffer.isBuffer(body) ? body : Buffer.from(body);
  const bodyHash = crypto.createHash('sha256').update(rawBody).digest('hex');
  return [
    SIGNATURE_VERSION,
    method.toUpperCase(),
    pathAndQuery,
    timestamp,
    nonce,
    bodyHash,
  ].join('\n');
}

function buildServiceHeaders(method, pathAndQuery, body = Buffer.alloc(0)) {
  const rawBody = Buffer.isBuffer(body) ? body : Buffer.from(body);
  const timestamp = String(Math.floor(Date.now() / 1000));
  const nonce = crypto.randomBytes(16).toString('hex');
  const canonical = canonicalServiceRequest(method, pathAndQuery, timestamp, nonce, rawBody);
  const signature = crypto
    .createHmac('sha256', config.internalServiceSecret)
    .update(canonical)
    .digest('hex');
  return {
    'X-Service-Version': SIGNATURE_VERSION,
    'X-Service-Timestamp': timestamp,
    'X-Service-Nonce': nonce,
    'X-Service-Signature': signature,
  };
}

module.exports = { buildServiceHeaders, canonicalServiceRequest };

File: backend/app/security/service_auth.py
import hashlib
import hmac
import time
import uuid

from ..config import INTERNAL_SERVICE_SECRET

SIGNATURE_VERSION = "v1"


def canonical_service_request(
    method: str,
    path_and_query: str,
    timestamp: str,
    nonce: str,
    body: bytes = b"",
) -> str:
    body_hash = hashlib.sha256(body).hexdigest()
    return "\n".join(
        [
            SIGNATURE_VERSION,
            method.upper(),
            path_and_query,
            timestamp,
            nonce,
            body_hash,
        ]
    )


def sign_service_request(canonical: str, secret: str = INTERNAL_SERVICE_SECRET) -> str:
    return hmac.new(secret.encode(), canonical.encode(), hashlib.sha256).hexdigest()


def build_service_headers(method: str, path_and_query: str, body: bytes = b"") -> dict[str, str]:
    timestamp = str(int(time.time()))
    nonce = uuid.uuid4().hex
    canonical = canonical_service_request(method, path_and_query, timestamp, nonce, body)
    signature = sign_service_request(canonical)
    return {
        "X-Service-Version": SIGNATURE_VERSION,
        "X-Service-Timestamp": timestamp,
        "X-Service-Nonce": nonce,
        "X-Service-Signature": signature,
    }
